From 39125fa7991fa85b0f8d4f51e7b5625d6f493475 Mon Sep 17 00:00:00 2001
From: Noriko Hosoi <nhosoi@redhat.com>
Date: Thu, 13 Jun 2013 18:22:21 -0700
Subject: [PATCH] Ticket #569 - examine replication code to reduce amount of
 stored state information

Description: commit c7f6f161f4967635d6f02b029be571d88ec61961
made this change:
  In case the deleted value list in an attribute is empty:
  * this means the entry is deleted and has no more attributes,
  * when writing the attr to disk we would loose the AD-csn.
  * Add an empty value to the set of deleted values. This will
  * never be seen by any client. It will never be moved to the
  * present values and is only used to preserve the AD-csn.
The AD-csn size was not counted for the buffer size to allocate.
This patch adds the size.

https://fedorahosted.org/389/ticket/569

Reviewed by Nathan (Thanks!!)

(cherry picked from commit ca02529c797903139c56bd9b8b90e8cd5bd0c3f4)

Conflicts:
	ldap/servers/slapd/entry.c

(cherry picked from commit 340b9dc5f979d75ab5bb92c7aa61808ff1fddd1f)
---
 ldap/servers/slapd/entry.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/ldap/servers/slapd/entry.c b/ldap/servers/slapd/entry.c
index 1341296..340844d 100644
--- a/ldap/servers/slapd/entry.c
+++ b/ldap/servers/slapd/entry.c
@@ -1476,7 +1476,17 @@ entry2str_internal_size_attrlist( const Slapi_Attr *attrlist, int entry2str_ctrl
 			/* ";adcsn-" + a->a_deletioncsn */
 			if ( a->a_deletioncsn )
 			{
-				elen+= 1 + LDIF_CSNPREFIX_MAXLENGTH + CSN_STRSIZE;
+				elen += 1 + LDIF_CSNPREFIX_MAXLENGTH + CSN_STRSIZE;
+			}
+			if ( valueset_isempty(&a->a_deleted_values)) {
+				/* this means the entry is deleted and has no more attributes,
+				 * when writing the attr to disk we would loose the AD-csn.
+				 * Add an empty value to the set of deleted values. This will 
+				 * never be seen by any client. It will never be moved to the 
+				 * present values and is only used to preserve the AD-csn
+				 * We need to add the size for that.
+				 */
+				elen += 1 + LDIF_CSNPREFIX_MAXLENGTH + CSN_STRSIZE;
 			}
 		}
 	}
-- 
1.9.3

